docs/source/cpp/io.rst

```diff
@@ -91,4 +101,42 @@ Concrete implementations are available for
 
   Tasks that use filesystems will typically run on the
   :ref:`I/O thread pool<io_thread_pool>`.  For filesystems that support high levels
-  of concurrency you may get a benefit from increasing the size of the I/O thread pool.
\ No newline at end of file
+  of concurrency you may get a benefit from increasing the size of the I/O thread pool.
+
+Defining new FileSystems
+========================
+
+Build complexity can be decreased by compartmentalizing a FileSystem
+implementation into a separate shared library, which applications may
+link or load dynamically. Arrow's built-in FileSystem implementations
+also follow this pattern. Before a scheme can be used with any of the
+`High-level factory functions`_ the library which contains it must be
```

Comment:
I don't think "High-level factory functions" will link to anything as there does not seem to be any such heading in this document. You may want to create a `ref` and corresponding target.